core/consensus/aura/src/lib.rs

```diff
@@ -314,3 +336,140 @@ pub fn import_queue<B, C>(config: Config, client: Arc<C>) -> impl ImportQueue<B>
+	use network::ProtocolConfig;
+	use parking_lot::Mutex;
+	use tokio::runtime::current_thread;
+	use keyring::Keyring;
+	use client::BlockchainEvents;
+	use primitives::H256;
+	use test_client;
+
+	type Error = client::error::Error;
+
+	type TestClient = client::Client<test_client::Backend, test_client::Executor, TestBlock>;
+
+	struct DummyFactory(Arc<TestClient>);
+	struct DummyProposer(u64, Arc<TestClient>);
+
+	impl Environment<TestBlock> for DummyFactory {
+		type Proposer = DummyProposer;
+		type Error = Error;
+
+		fn init(&self, parent_header: &<TestBlock as BlockT>::Header, _authorities: &[AuthorityId], _sign_with: Arc<ed25519::Pair>)
+			-> Result<DummyProposer, Error>
+		{
+			Ok(DummyProposer(parent_header.number + 1, self.0.clone()))
+		}
+	}
+
+	impl Proposer<TestBlock> for DummyProposer {
+		type Error = Error;
+		type Create = Result<TestBlock, Error>;
+		type Evaluate = Result<bool, Error>;
+
+		fn propose(&self) -> Result<TestBlock, Error> {
+			self.1.new_block().unwrap().bake().map_err(|e| e.into())
+		}
+
+		fn evaluate(&self, proposal: &TestBlock) -> Result<bool, Error> {
+			Ok(proposal.header.number == self.0)
+		}
+	}
+
+	#[derive(Encode, Decode)]
+	pub struct AuraSeal(u64, ed25519::Signature);
+
+	impl CompatibleDigestItem for runtime_primitives::generic::DigestItem<primitives::H256, u64> {
+		/// Construct a digest item which is a slot number and a signature on the
+		/// hash.
+		fn aura_seal(slot_number: u64, signature: ed25519::Signature) -> Self {
+			runtime_primitives::generic::DigestItem::Seal(slot_number, signature)
+		}
+
+		/// If this item is an Aura seal, return the slot number and signature.
+		fn as_aura_seal(&self) -> Option<(u64, &ed25519::Signature)> {
+			match self {
+				runtime_primitives::generic::DigestItem::Seal(slot, ref sign) => Some((*slot, sign)),
+				_ => None
+			}
+		}
+
+	}
+
+
+	const SLOT_DURATION: u64 = 1;
+	const TEST_ROUTING_INTERVAL: Duration = Duration::from_millis(50);
+
+	#[test]
+	fn authoring_blocks() {
+		let net = TestNet::new(3);
+
+		let peers = &[
+			(0, Keyring::Alice),
+			(1, Keyring::Bob),
+			(2, Keyring::Charlie),
+		];
+
+		let voters: Vec<_> = peers.iter()
+			.map(|&(_, ref key)| AuthorityId(key.to_raw_public()))
+			.collect();
+
+		// set up peers
+		// for (peer_id, key) in peers {
+		// 	let config = Config {
+		// 			local_key: Some(key.clone()),
+		// 			slot_duration: SLOT_DURATION
+		// 		};
+		// 	net.add_peer_with_verifier(ProtocolConfig {
+		// 		roles: Roles::AUTHORITY
+		// 	}, |client| Verifier{ client, config });
+		// }
+
+
+		let net = Arc::new(Mutex::new(net));
+		let mut import_notifications = Vec::new();
+
+		let mut runtime = current_thread::Runtime::new().unwrap();
+		for (peer_id, key) in peers {
+			let mut client = net.lock().peer(*peer_id).client().clone();
+			let environ = Arc::new(DummyFactory(client.clone()));
+			import_notifications.push(
+				client.import_notification_stream()
+					.take_while(|n| Ok(n.header.number() < &10))
```

Comment:
it may be possible for this test to pass without each node getting the others' blocks. Perhaps we could add another non-authority node to the network